streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1063,4 +1072,49 @@ object ZSink extends ZSinkPlatformSpecific {
       def stepPure(state: State, a: Any): Step[State, Nothing] = Step.done(state, Chunk.empty)
       def extractPure(state: State): Either[Nothing, B]        = Right(b)
     }
+
+  /**
+   * Creates a sink which throttles input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` effectful
+   * function. Stream elements are mapped to `Option[A]`, and `None` denotes that a given element has
+   * been throttled.
+   */
+  final def throttle[R, E, A](units: Long, duration: Duration)(
+    costFn: A => ZIO[R, E, Long]
+  ): ZManaged[R with Clock, E, ZSink[R with Clock, E, Nothing, A, Option[A]]] = {
```

Comment:
So this doesn't need to be a `Managed` now right? The `Ref` and clock read can be done in `initial`.

Reply:
I forgot to change this. I wanted to send some code in before leaving for a festival. Working on this PR right now.

Reply:
I have some trouble removing the `ZManaged` from the type signature. For each stream element, the `ZSink` gets reset (by calling `initial` state), which is why we need to keep a `Ref` as the state holder. Also, the ref initialization needs to happen exactly once, which means that we cannot make the ref in `initial`.

Reply:
And since `initial` doesn't take any arguments, there is no way to "inject" the old state.

Reply:
Right, I see what you mean. Let's go ahead and merge this then.